src/core/server/saved_objects/export/find_with_point_in_time.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Logger } from '../../logging';
+import { SavedObjectsClientContract, SavedObjectsFindOptions } from '../types';
+import { SavedObjectsFindResponse } from '../service';
+
+/**
+ * Returns a generator to help page through large sets of saved objects.
+ *
+ * The generator wraps calls to `SavedObjects.find` and iterates over
+ * multiple pages of results using `_pit` and `search_after`. This will
+ * open a new Point In Time (PIT), and continue paging until a set of
+ * results is received that's smaller than the designated `perPage`.
+ *
+ * Once you have retrieved all of the results you need, it is recommended
+ * to call `close()` to clean up the PIT and prevent Elasticsearch from
+ * consuming resources unnecessarily. This will automatically be done for
+ * you if you reach the last page of results.
+ *
+ * @example
+ * ```ts
+ * const finder = findWithPointInTime({
+ *   logger,
+ *   savedObjectsClient,
+ * });
+ *
+ * const options: SavedObjectsFindOptions = {
+ *   type: 'visualization',
+ *   search: 'foo*',
+ *   perPage: 100,
+ * };
+ *
+ * const responses: SavedObjectFindResponse[] = [];
+ * for await (const response of finder.find(options)) {
+ *   responses.push(...response);
+ *   if (doneSearching) {
+ *     await finder.close();
+ *   }
+ * }
+ * ```
+ */
+export function findWithPointInTime({
+  logger,
+  savedObjectsClient,
+}: {
+  logger: Logger;
+  savedObjectsClient: SavedObjectsClientContract;
+}) {
+  return new FindWithPointInTime({ logger, savedObjectsClient });
+}
+
+/**
+ * @internal
+ */
+export class FindWithPointInTime {
+  readonly #log: Logger;
+  readonly #savedObjectsClient: SavedObjectsClientContract;
+  #open?: boolean;
+  #perPage?: number;
+  #pitId?: string;
+  #type?: string | string[];
+
+  constructor({
+    savedObjectsClient,
+    logger,
+  }: {
+    savedObjectsClient: SavedObjectsClientContract;
+    logger: Logger;
+  }) {
+    this.#log = logger;
+    this.#savedObjectsClient = savedObjectsClient;
+  }
+
+  async *find(options: SavedObjectsFindOptions) {
+    this.#open = true;
+    this.#type = options.type;
+    // Default to 1000 items per page as a tradeoff between
+    // speed and memory consumption.
+    this.#perPage = options.perPage ?? 1000;
+
```

Comment:
Oh, great point - that wasn't the intended usage (an instance of the `finder` isn't designed to be reused).

This can probably be avoided by removing the options from `find` altogether and instead putting them in the constructor, so you can only provide them once per instance. Will work on updating that

```ts
    const findOptions: SavedObjectsFindOptions = {
      type: types,
      hasReference,
      hasReferenceOperator: hasReference ? 'OR' : undefined,
      search,
      namespaces: namespace ? [namespace] : undefined,
    };

    const finder = createPointInTimeFinder({
      logger: this.#log,
      savedObjectsClient: this.#savedObjectsClient,
      findOptions,
    });

    const hits: SavedObjectsFindResult[] = [];
    for await (const result of finder.find()) {
      hits.push(...result.saved_objects);
    }
```

Reply:
Moving the options to the constructor seems good, but I don't think it would totally avoid unexpected behavior when executing my code snippet (as multiple pit ids will be created, and only the last one used)

I think (in addition to what you suggest) we could have it throw if `find` is called multiple times (or if it is called again before `close` has been called)

Reply:
> have it throw if find is called multiple times (or if it is called again before close has been called)

SGTM, will add this.